File: Vitis-AI-Quantizer/vai_q_tensorflow/tensorflow/lite/delegates/gpu/cl/kernels/prelu.cc
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/lite/delegates/gpu/cl/kernels/prelu.h"

#include "absl/strings/str_cat.h"
#include "absl/types/variant.h"
#include "tensorflow/lite/delegates/gpu/cl/kernels/util.h"
#include "tensorflow/lite/delegates/gpu/common/tensor.h"

namespace tflite {
namespace gpu {
namespace cl {

PReLU::PReLU(const OperationDef& definition, const PReLUAttributes& attr)
    : ElementwiseOperation(definition) {
  if (attr.clip != 0) {
    clip_ = FLT(definition.precision, attr.clip);
  }
}

PReLU::PReLU(PReLU&& operation)
    : ElementwiseOperation(std::move(operation)),
      clip_(std::move(operation.clip_)),
      alpha_(std::move(operation.alpha_)) {}

PReLU& PReLU::operator=(PReLU&& operation) {
  if (this != &operation) {
    clip_ = std::move(operation.clip_);
    alpha_ = std::move(operation.alpha_);
    ElementwiseOperation::operator=(std::move(operation));
  }
  return *this;
}

void PReLU::SetLinkIndex(int index) {
  clip_.SetName(absl::StrCat("prelu_clip", index));
  alpha_.SetName(absl::StrCat("prelu_alpha_", index));
}

std::string PReLU::GetCoreCode(const std::string& src,
                               const std::string& z_coord,
                               const std::string& address) const {
  if (!clip_.Active()) {
    return absl::StrCat(src, " = max((FLT4)(0.0f), ", src,
                        ") + min((FLT4)(0.0f), ", src, ") * ",
                        alpha_.ReadLinearFLT4(z_coord), ";\n");
  } else {
    return absl::StrCat(src, " = clamp(", src, ", (FLT4)(0.0f), (FLT4)(",
                        clip_.GetName(), ")) + min((FLT4)(0.0f), ", src, ") * ",
                        alpha_.ReadLinearFLT4(z_coord), ";\n");
  }
}

std::string PReLU::GetArgsDeclaration() const {
  std::string args = absl::StrCat(",\n    ", alpha_.GetDeclaration());
  if (clip_.Active()) {
    args = absl::StrCat(args, ",\n    ", clip_.GetDeclaration());
  }
  return args;
}

Status PReLU::BindArguments(CLKernel* kernel) {
  RETURN_IF_ERROR(kernel->SetMemoryAuto(alpha_.GetMemoryPtr()));
  if (clip_.Active()) {
    RETURN_IF_ERROR(kernel->SetBytesAuto(clip_));
  }
  return OkStatus();
}

Status CreatePReLU(const CreationContext& creation_context,
                   const OperationDef& definition, const PReLUAttributes& attr,
                   PReLU* result) {
  auto alpha = absl::get_if<::tflite::gpu::Tensor<Linear, DataType::FLOAT32>>(
      &attr.alpha);
  if (!alpha) {
    return InvalidArgumentError("Alpha is missing");
  }
  *result = PReLU(definition, attr);
  RETURN_IF_ERROR(result->UploadParameters(*alpha, creation_context.context));
  result->SetLinkIndex(0);
  return OkStatus();
}

}  // namespace cl
}  // namespace gpu
}  // namespace tflite

File: src/vai_optimizer/pytorch_binding/pytorch_nndct/nn/modules/prelu.py


#
# Copyright 2019 Xilinx Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import torch
from torch.autograd import Variable
import math

from nndct_shared.utils import NndctOption, NndctScreenLogger
from nndct_shared.quantization import maybe_get_quantizer
from nndct_shared.quantization import quantize_tensors 
from .quant_noise import eval_qnoise
import pytorch_nndct.utils as py_utils
import torch.nn.functional as F

__all__ = ['PReLU']

class deephi_PReLU(torch.nn.PReLU):
  r"""DeePhi PReLU operation, support float and double"""

  def __init__(self, *args, **kwards):
    super(deephi_PReLU, self).__init__(*args, **kwards)
    self.params_name = None
    self.node = None
    self.quant_mode, self.quantizer = maybe_get_quantizer()
    self.param_saved = False
    self.param_quantized = False

  def forward(self, input):
    # quantize parameters
    qweight = None
    inplace = (NndctOption.nndct_quant_off.value or 
        self.quantizer is not None and self.quantizer.inplace)
    if (not self.param_quantized):
      if inplace:
        _ = quantize_tensors(
            [self.weight],
            self.node,
            tensor_names = [self.params_name[0]],
            tensor_type = 'param')[0]
        qweight = self.weight
      else:
        qweight = quantize_tensors(
            [self.weight],
            self.node,
            tensor_names = [self.params_name[0]],
            tensor_type = 'param')[0]
      if not NndctOption.nndct_quant_off.value:
        self.param_quantized = True
    else:
      qweight = self.weight

    # quantize input tensor
    qinput = quantize_tensors([input], self.node, tensor_type='input')[0]
    output = torch.nn.functional.prelu(input = qinput, weight = qweight) 
    output = quantize_tensors([output], self.node)[0]

    return output

@py_utils.register_quant_op
def PReLU(*args, **kwargs):
  quant_mode, _ = maybe_get_quantizer()
  if quant_mode == None:
    return torch.nn.PReLU(*args, **kwargs)
  return deephi_PReLU(*args, **kwargs)
